File: 5.InfixtoPostfix/infix_postfix.py
def precedence(op):
    if op == '^':
        return 3
    if op == '*' or op == '/':
        return 2
    if op == '+' or op == '-':
        return 1
    return 0

def infix_to_postfix(expr):
    st = []
    result = []

    for ch in expr:
        if ch.isalnum():  # letter or digit
            result.append(ch)
        elif ch == '(':
            st.append(ch)
        elif ch == ')':
            while st and st[-1] != '(':
                result.append(st.pop())
            st.pop()  # remove '('
        else:
            # normal operator
            while (st and precedence(st[-1]) >= precedence(ch) and ch != '^'):
                result.append(st.pop())
            st.append(ch)

    while st:
        result.append(st.pop())

    return " ".join(result)


# main
expr = input().strip()
print(infix_to_postfix(expr))

File: 5.InfixtoPostfix/infix_postfix.java
// Write your code here!!

import java.util.*;

public class InfixToPostfix {
    static int precedence(char op) {
        if (op == '^') return 3;
        if (op == '*' || op == '/') return 2;
        if (op == '+' || op == '-') return 1;
        return 0;
    }

    static String infixToPostfix(String expr) {
        Stack<Character> st = new Stack<>();
        StringBuilder result = new StringBuilder();

        for (char ch : expr.toCharArray()) {
            if (Character.isLetterOrDigit(ch)) {
                result.append(ch).append(' ');
            } else if (ch == '(') {
                st.push(ch);
            } else if (ch == ')') {
                while (!st.isEmpty() && st.peek() != '(')
                    result.append(st.pop()).append(' ');
                st.pop();
            } else {
                while (!st.isEmpty() && precedence(st.peek()) >= precedence(ch) && ch != '^')
                    result.append(st.pop()).append(' ');
                st.push(ch);
            }
        }

        while (!st.isEmpty())
            result.append(st.pop()).append(' ');

        return result.toString();
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String expr = sc.next();
        System.out.println(infixToPostfix(expr));
    }
}
